### ISSUE TITLE
Tuple return values for generics methods without type constraints should trigger CS8653

### ISSUE THREAD

**Issue by fredjeronimo** [dotnet/roslyn#42844]
**Version Used**: 
```
>csc /?
Microsoft (R) Visual C# Compiler version 3.4.1-beta4-20120-08 (fa984e02)
```

**Steps to Reproduce**:

Build this program using generics methods:

```
#nullable enable
internal static class Program
{
    private static T TriggersCs8653<T>()
    {
        return default;
    }

    private static (int, T) DoesNotTriggerCs8653<T>()
    {
        return (1, default);
    }

    private static void Main()
    {
    }
}
```

**Actual Behavior**:

Produces a single build warning on line 6: `Program.cs(6,16,6,23): warning CS8653: A default expression introduces a null value when 'T' is a non-nullable reference type.`

Line 11 issues no warning.

**Expected Behavior**:

Behavior seems inconsistent. Tuple return values for generics methods without type constraints should also issue CS8653.

**Comment by jcouv** [comment]
Thanks for reporting this.

Investigated the issue and I think it will be easier to fix once we implement `T??` (issue https://github.com/dotnet/roslyn/issues/29146). Then `NullableWalker.VisitTupleExpression` can return an `(int, T??)` as the result type for `(int, default(T))` and we can produce the expected diagnostic from `VisitConversion` (`ImplicitTupleLiteral` case).

``` csharp
        [Fact, WorkItem(42844, "https://github.com/dotnet/roslyn/issues/42844")]
        public void ReturningValues_TupleWithDefaultUnconstrainedTypeParameter()
        {
            var source = @"
public class C<T>
{
    (int, T) M()
    {
        return (0, default); //
    }
    (int, T) M2()
    {
        (int, T) t = (0, default); //
        return t;
    }
    (string, int) M3()
    {
        return (null, 0); // 3
    }
    (string, int) M4()
    {
        (string, int) t = (null, 0); // 4
        return t;
    }
}";

            var comp = CreateNullableCompilation(source);
            comp.VerifyDiagnostics(
                // (15,16): warning CS8619: Nullability of reference types in value of type '(string?, int)' doesn't match target type '(string, int)'.
                //         return (null, 0); // 3
                Diagnostic(ErrorCode.WRN_NullabilityMismatchInAssignment, "(null, 0)").WithArguments("(string?, int)", "(string, int)").WithLocation(15, 16),
                // (19,27): warning CS8619: Nullability of reference types in value of type '(string?, int)' doesn't match target type '(string, int)'.
                //         (string, int) t = (null, 0); // 4
                Diagnostic(ErrorCode.WRN_NullabilityMismatchInAssignment, "(null, 0)").WithArguments("(string?, int)", "(string, int)").WithLocation(19, 27)
                );
        }
```

**Comment by jaredpar** [comment]
This correctly produces a warning now that we've implemented defaultable generics

https://sharplab.io/#v2:EYLgtghgzgLgpgJwD4GIB2BXANliwtwAEcaeBAsAFACWa8CpWhAAgIwBsLATC6wOxUA3lUKjCABwTUAbhHi9OAFUKKpAczWIoAYSgAOdgFYAzAB5FAPgAUAShFjhlMc5Z9CAEzgAzCNhgBue1EAXyogiSlZeTZOK1oYABoVG0IAEQB7OCgAOXSYVWoNRF0DE3NrOycHcOdmNytWJM8fPxtAqpCwjoiZOSIYlgAWQgBZCFpbcMdnUMpgoA===

Moving to test only issue since we still need to add a regression test here.

**Comment by fredjeronimo** [comment]
That's great. Thanks for chasing this and for the update.
